### PR TITLE
Optimize tracing loading

### LINKED ISSUE #1499: Silent failure if bytecode instrumentation is missing all expected methods
This is similar to https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/issues/1469 but in this case the instrumentation finds the type but not the expected method.

### LINKED ISSUE #1505: Loading unnecessary exporter modules when instrumenting app
When running a simple test like StrongNamedTests the logs show that many unexpected exporter modules are loaded, e.g.: Jaeger. This is issue is to investigate why and fix, if appropriate, any exporter module that is being loaded by not actually used.

List for the modules loaded for the StrongNamedTests when running on net462:
```terminal
Anonymously Hosted DynamicMethods Assembly
CrowdStrike.Sensor.ScriptControl
Google.Protobuf
Grpc.Core
Grpc.Core.Api
Microsoft.Extensions.Logging.Abstractions
mscorlib
netstandard
OpenTelemetry
OpenTelemetry.Api
OpenTelemetry.AutoInstrumentation
OpenTelemetry.AutoInstrumentation.Loader
OpenTelemetry.Exporter.Console
OpenTelemetry.Exporter.Jaeger
OpenTelemetry.Exporter.OpenTelemetryProtocol
OpenTelemetry.Exporter.Prometheus
OpenTelemetry.Exporter.Zipkin
OpenTelemetry.Instrumentation.AspNet
OpenTelemetry.Instrumentation.AspNet.TelemetryHttpModule
OpenTelemetry.Instrumentation.GrpcNetClient
OpenTelemetry.Instrumentation.Http
OpenTelemetry.Instrumentation.Process
OpenTelemetry.Instrumentation.Runtime
OpenTelemetry.Instrumentation.SqlClient
OpenTelemetry.Instrumentation.Wcf
OpenTelemetry.Shims.OpenTracing
OpenTracing
System
System.Buffers
System.Configuration
System.Core
System.Diagnostics.DiagnosticSource
System.Memory
System.Net.Http
System.Runtime.CompilerServices.Unsafe
System.ServiceModel
System.Web
System.Xml
TestApplication.StrongNamed
TestLibrary.InstrumentationTarget
```

### PR DESCRIPTION
**Pull request by RassK**
## Why

Partially fixes #1505 

## What

This PR fixes only tracing related parts. metrics and logs may need additional work.

Wraps external extension methods with internal method to prevent loading the library if it's not needed.
Seems .NET Core compiler does not inline it, but .NET Fx is too aggressive, so inlining had to be explicitly disabled.

## Tests

Added minimal loading list

## Checklist

- [x] New features are covered by tests.

### CODE REVIEW

File: test/IntegrationTests/ModuleTests.Default.NetCore.verified.txt [hunk 1]
```diff
@@ -5,10 +5,8 @@
   OpenTelemetry.AutoInstrumentation.Loader,
   OpenTelemetry.AutoInstrumentation.StartupHook,
   OpenTelemetry.Exporter.Console,
```
Comment (pjanotti) [inline test/IntegrationTests/ModuleTests.Default.NetCore.verified.txt:7]: Starting the review here - I may see why later, but, do you know why `Console` and `Prometheus` are still being loaded?

Comment (pjanotti) [reply test/IntegrationTests/ModuleTests.Default.NetCore.verified.txt:7]: > This PR fixes only tracing related parts

Yes, these are coming via Metrics.

File: src/OpenTelemetry.AutoInstrumentation/Configuration/EnvironmentConfigurationTracerHelper.cs [hunk 2]
```diff
@@ -63,57 +56,107 @@ public static TracerProviderBuilder UseEnvironmentVariables(this TracerProviderB
...
+            TracesExporter.Zipkin => Wrappers.AddZipkinExporter(builder),
+            TracesExporter.Jaeger => Wrappers.AddJaegerExporter(builder),
+            TracesExporter.Otlp => Wrappers.AddOtlpExporter(builder, settings),
+            TracesExporter.None => builder,
+            _ => throw new ArgumentOutOfRangeException($"Traces exporter '{settings.TracesExporter}' is incorrect")
+        };
+    }
+
+    /// <summary>
+    /// This class wraps external extensions methods to ensure the dlls are not loaded, if not necessary.
```
Comment (pjanotti) [inline src/OpenTelemetry.AutoInstrumentation/Configuration/EnvironmentConfigurationTracerHelper.cs:77]: It seems worth adding a mention that preventing inlining is required.

File: src/OpenTelemetry.AutoInstrumentation/Configuration/EnvironmentConfigurationTracerHelper.cs [hunk 3]
```diff
@@ -63,57 +56,107 @@ public static TracerProviderBuilder UseEnvironmentVariables(this TracerProviderB
...
+    /// </summary>
+    private static class Wrappers
+    {
+        // Invstrumentations
```
Comment (pjanotti) [inline src/OpenTelemetry.AutoInstrumentation/Configuration/EnvironmentConfigurationTracerHelper.cs:81]: typo

**Review by pjanotti — APPROVED** [review]

File: src/OpenTelemetry.AutoInstrumentation/Configuration/EnvironmentConfigurationTracerHelper.cs (hunk 2 above)
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Configuration/EnvironmentConfigurationTracerHelper.cs]: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/1530/commits/734358f34611e6883b71e159b431320518a5e3cd

**Review by pjanotti — APPROVED** [review]

**Comment by pellared** [comment]
Nice! 🎉 👍

**Pull request merged**